### PR TITLE
Sprint 43

### PR DESCRIPTION
**Pull request by ericsorenson**
# Sprint 43

### Overview
Due to various issues we have encountered with using the serverless framework (AWS CloudFormation), we have
decided to migrate as many stacks as we can to instead be deployed via terraform. The approach we are taking
is to have a single lambda which is running an express service which will map the various api gateway requests to the
corresponding endpoints. There is an aws-serverless-express plugin which maps the api gateway events into express
request events.The benefits to having a single lambda hosting all of our code include:
- faster deploys
- limited amount of cold starts
- less infrastructure overhead (single file having all our routes vs 15 serverless files)
- single express service to run locally when doing development## What is Changing?
In the short term, we are removing as many serverless.yml files as possible. All of those \*-serverless.yml files were responsible for deploying our api endpoints / stacks. We added two directories: web-api/terraform/template/api and web-api/terraform/template/api-public. They both contain a single index.js which imports the necessary express application (web-api/src/app.js and web-api/src/app-public.js) and runs parcel over them to create a bundled .js file used for deployment. Parcel will generate a ./dist directory inside those directories which is used by terraform to deploy the lambdas.
Inside the web-api/terraform folder, we are creating some new \*.tf files: api.tf and api-public.tf. Those two files will contain all of the aws resources needed for hosting our web api, including the api gateway, route53 records, custom api gateway domain name, base path mapping, and the lambda, etc.  There are a couple of other directories inside the terraform folder for one off lambdas, such as cron job lambdas, or the dynamodb streams lambda.
For running locally, modified the run-local.sh script to invoke nodemon on our app-local.js and app-public-local.js. These two new .js files are used for hosting the express application locally on ports 4000 and 5000, respectively.## What do you need to do?
- You will need to re-run the ./iam/terraform/enviroment-specific terraform for each environment. We added a new role / policy which is needed for the api gateway to communicate with the lambdas.
- You will want to manually (or using parts of the cleanup script) delete any of the old cloud formation stacks, api gateways, and custom gateway domain names, and route53 records.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -566,38 +372,35 @@ jobs:
       - run:
           name: 'Deploy - Web API - Smoke Tests - us-east-1'
           command: |
-            if [ "${CIRCLE_BRANCH}" == "staging" ] || [ "${CIRCLE_BRANCH}" == "irs" ]; then
+            if [ "${CIRCLE_BRANCH}" == "develop" ] || [ "${CIRCLE_BRANCH}" == "experimental1" ] || [ "${CIRCLE_BRANCH}" == "experimental2" ] || [ "${CIRCLE_BRANCH}" == "staging" ] || [ "${CIRCLE_BRANCH}" == "irs" ]; then
```
Comment (julialeague) [inline .circleci/config.yml:375]: I think we can change this on the ustaxcourt side to fit our environments - just want to note that so that I can follow up later.

File: clear-env.sh [hunk 2]
```diff
@@ -1,25 +1,35 @@
 #!/bin/bash
 
-# usage:
+# Usage
 #   ./clear-env.sh $ENV
```
Comment (julialeague) [inline clear-env.sh:4]: It would be nice to have a small description of what the script does too.

**Comment by kfoley-18F** [comment]
We use this list when performing a code review to ensure that all tasks have been completed.

- [ ] begin by deleting any existing copy of the Git repository, running `docker system prune` to remove any cached artifacts from any prior code review
- [ ] clone a fresh copy of the Git repository (`git clone https://github.com/ustaxcourt/ef-cms.git`), fetch the pull request and name it for the sprint (e.g., `git fetch origin pull/204/head:sprint-43`), and then switch to that branch (e.g. `git checkout sprint-43`)
- [ ] review the pull request itself, to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] stand up the site locally, with `./docker-run.sh`
	- [ ] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
		- [ ] for internal Court functionality, perform the most thorough testing in Chrome, though also test in Edge and Firefox
		- [ ] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/), including Internet Explorer 11
		- [ ] test in Mobile Safari and Mobile Chrome, with the caveat that not all internal Court functionality will be necessary on these platforms
	- [ ] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [ ] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [ ] review for accessibility
		- [ ] use an automated audit tool, such as Chrome Audit or aXe
		- [ ] navigate site only with the keyboard
		- [ ] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [ ] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [ ] [review static code analysis results](https://sonarcloud.io/organizations/flexion-github/projects) in the vendor’s SonarCloud environment
- [ ] examine OWASP ZAP output in `docs/`, to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [ ] skim all new code, in the context of existing code, [looking for problems](#what-we-look-for) (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] look at code coverage of tests in [the vendor's SonarCloud environment](https://sonarcloud.io/organizations/flexion-github/projects)
	- [ ] methodically review all new tests for correctness, quality of naming
- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

File: cypress-smoketests/fixtures/caseMigrations.js [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+exports.BASE_CASE = {
+  ...MOCK_CASE,
+  associatedJudge: 'Chief Judge',
+  caseCaption: 'A Migrated Casee',
+  caseId: undefined,
+  preferredTrialCity: 'Washington, District of Columbia',
+  status: 'Calendared',
+};
+
+exports.CASE_WITH_OTHER_PETITIONERS = {
```
Comment (julialeague) [inline cypress-smoketests/fixtures/caseMigrations.js:13]: @ericsorenson Not a huge deal, but I see potential to DRY up this and `CASE_WITH_OTHER_FILERS` below.

File: cypress-smoketests/fixtures/example.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
+  "body": "Fixtures are a great way to mock data for responses to routes"
+}
```
Comment (julialeague) [inline cypress-smoketests/fixtures/example.json:5]: @ericsorenson newline

File: cypress-smoketests/integration/case-creation.spec.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  fillInCreateCaseFromPaperForm,
+} = require('../../cypress/support/pages/create-paper-petition');
+const { getUserToken, login } = require('../support/pages/login');
+const { goToMyDocumentQC } = require('../support/pages/document-qc');
+
+let token = null;
+
+describe('Petitioner', () => {
+  before(async () => {
+    const results = await getUserToken(
```
Comment (julialeague) [inline cypress-smoketests/integration/case-creation.spec.js:34]: @ericsorenson I think there's been some talk about figuring out how we can stop relying on these predetermined Cognito users. Do you know if there's a card in the backlog somewhere about this?

Comment (julialeague) [reply cypress-smoketests/integration/case-creation.spec.js:34]: We talked about this in our last launch planning meeting, and there's a card created for that in the backlog 👍

File: cypress-smoketests/integration/filings-and-qc.spec.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  goToSelectDocumentType,
+  selectDocumentType,
+  submitDocument,
+  uploadDocumentFile,
+} = require('../support/pages/petitioner-dashboard');
+const { getUserToken, login } = require('../support/pages/login');
+
+let token = null;
+
+describe.skip('Petitions clerk', () => {
```
Comment (julialeague) [inline cypress-smoketests/integration/filings-and-qc.spec.js:28]: @ericsorenson Just wanted to check if we have something logged for updating these so they're functioning/not skipped?

File: cypress-smoketests/support/pages/login.js [hunk 7]
```diff
@@ -51,27 +53,28 @@ const getUserToken = async (username, password) => {
...
-    cy.get('#advanced-search-button').click();
-    cy.get('#no-search-results').should('exist');
-  });
-});
+  const services = apis
+    .filter(api => api.name.includes(`gateway_api_${ENV}`))
+    .reduce((obj, api) => {
+      obj[
+        api.name.replace(`_${ENV}`, '')
+      ] = `https://${api.id}.execute-api.${awsRegion}.amazonaws.com/${ENV}`;
```
Comment (julialeague) [inline cypress-smoketests/support/pages/login.js:75]: @ericsorenson I feel like there's potential here to make the whole url configurable.

File: iam/terraform/environment-specific/main/api-gateway-authorizer.tf [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  "Statement": [
+    {
+      "Action": "lambda:InvokeFunction",
+      "Effect": "Allow",
+      "Resource": "*"
+    }
+  ]
+}
+EOF
+}
```
Comment (julialeague) [inline iam/terraform/environment-specific/main/api-gateway-authorizer.tf:37]: @ericsorenson newline

File: shared/src/business/entities/DocketRecord.test.js [hunk 9]
```diff
@@ -121,5 +121,65 @@ describe('DocketRecord', () => {
...
+
+    it('passes validation when isLegacy is true and isStricken is defined', () => {
+      const invalidDocketRecord = new DocketRecord(
+        {
+          description: 'Test Docket Record',
+          eventCode: 'O',
+          filingDate: new Date('2000-01-01').toISOString(),
+          index: 0,
+          isLegacy: true,
+          isStriken: false,
```
Comment (julialeague) [inline shared/src/business/entities/DocketRecord.test.js:148]: Typo?

File: shared/src/business/entities/DocketRecord.test.js [hunk 10]
```diff
@@ -121,5 +121,65 @@ describe('DocketRecord', () => {
...
+        },
+        { applicationContext },
+      );
+
+      expect(invalidDocketRecord.isValid()).toBeFalsy();
```
Comment (julialeague) [inline shared/src/business/entities/DocketRecord.test.js:153]: @ericsorenson The test description doesn't match the expectation here, it seems.

File: shared/src/business/entities/Document.test.js [hunk 11]
```diff
@@ -205,6 +214,444 @@ describe('Document entity', () => {
...
+          { applicationContext },
+        );
+        expect(document.isValid()).toBeFalsy();
+        expect(document.getFormattedValidationErrors()).toMatchObject({
+          isLegacy: '"isLegacy" is required',
+          isSealed: '"isSealed" is required',
+        });
+      });
+
+      it('should pass validation when "isLegacy" is true, "isSealed" and "isLegacy" are true', () => {
```
Comment (julialeague) [inline shared/src/business/entities/Document.test.js:248]: @ericsorenson just a typo or something, we could take out the first `when "isLegacy" is true`

File: shared/src/business/entities/Document.test.js [hunk 12]
```diff
@@ -205,6 +214,444 @@ describe('Document entity', () => {
...
+            isOrder: true,
+            isSealed: true,
+            secondaryDate: '2019-03-01T21:40:46.415Z',
+          },
+          { applicationContext },
+        );
+        expect(document.isValid()).toBeTruthy();
+      });
+
+      it('should pass validation when "isLegacy" is false, "isSealed" and "isLegacy" are undefined', () => {
```
Comment (julialeague) [inline shared/src/business/entities/Document.test.js:267]: @ericsorenson Same as above: https://github.com/ustaxcourt/ef-cms/pull/280/files#r450945304

File: shared/src/business/entities/Document.test.js [hunk 13]
```diff
@@ -205,6 +214,444 @@ describe('Document entity', () => {
...
+                },
+                { applicationContext },
+              );
+
+              expect(document.filedBy).toBeUndefined();
+              expect(document.isValid()).toBeTruthy();
+            });
+
+            it('should pass validation when "isAutoGenerated" is undefined and "filedBy" is undefined', () => {
+              // then: joi.when('isAutoGenerated'  is: false,
```
Comment (julialeague) [inline shared/src/business/entities/Document.test.js:369]: I'm not clear on the intention behind this type of comment - what's its purpose? I see it a few times in this test file.

File: web-api/src/middleware/apiGatewayHelper.test.js [hunk 14]
```diff
@@ -313,6 +289,18 @@ describe('getAuthHeader', () => {
...
+  it('should not return the user token and should throw an error if token is not present in query or anywhere else', () => {
+    let error;
+    let response;
+    try {
+      response = getAuthHeader({});
+    } catch (err) {
+      error = err;
+    }
+    expect(response).toEqual(null);
+    expect(error).toBeUndefined();
```
Comment (julialeague) [inline web-api/src/middleware/apiGatewayHelper.test.js:302]: @ericsorenson I'm wondering if this should actually have a defined error in this case where the AuthHeader isn't passed - the test description indicates an error should be thrown.

File: web-client/integration-tests/journey/petitionsClerkAddsNoticeToCase.js [hunk 15]
```diff
@@ -43,6 +43,13 @@ export const petitionsClerkAddsNoticeToCase = test => {
     expect(test.getState('validationErrors')).toEqual({});
     expect(test.getState('pdfPreviewUrl')).toBeDefined();
 
+    // await test.runSequence('setPDFSignatureDataSequence', {
```
Comment (julialeague) [inline web-client/integration-tests/journey/petitionsClerkAddsNoticeToCase.js:46]: @ericsorenson Why is this commented out?

File: web-client/src/presenter/actions/navigateToMessageDetailAction.test.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+  beforeAll(() => {
+    routeStub = jest.fn();
+
+    presenter.providers.router = {
+      route: routeStub,
+    };
+  });
+
+  it('navigates to message detail page without parent message id', async () => {
```
Comment (julialeague) [inline web-client/src/presenter/actions/navigateToMessageDetailAction.test.js:18]: @ericsorenson `without parent message id` makes me think the path param for parent message id shouldn't be passed into the url, but it looks like it is?

**Comment by julialeague** [comment]
@ericsorenson could we get this documentation updated to reflect the move away from multiple CloudFormation stacks? https://github.com/ustaxcourt/ef-cms/blob/1e1d7bdd66d654f5830b8b6a2d4bb3136d1bab7b/docs/CHECKLISTS.md#developer-check-lists

File: shared/src/business/entities/Document.test.js (hunk 13 above)
Comment (ericsorenson) [reply shared/src/business/entities/Document.test.js:369]: These were notes from building the tests, I'll remove them.

File: clear-env.sh (hunk 2 above)
Comment (ericsorenson) [reply clear-env.sh:6]: Updated!

File: cypress-smoketests/fixtures/example.json (hunk 4 above)
Comment (ericsorenson) [reply cypress-smoketests/fixtures/example.json]: Updated!

File: iam/terraform/environment-specific/main/api-gateway-authorizer.tf (hunk 8 above)
Comment (ericsorenson) [reply iam/terraform/environment-specific/main/api-gateway-authorizer.tf]: Updated!

File: shared/src/business/entities/DocketRecord.test.js (hunk 9 above)
Comment (ericsorenson) [reply shared/src/business/entities/DocketRecord.test.js]: Updated!

File: shared/src/business/entities/DocketRecord.test.js (hunk 10 above)
Comment (ericsorenson) [reply shared/src/business/entities/DocketRecord.test.js]: Updated!

File: shared/src/business/entities/Document.test.js (hunk 11 above)
Comment (ericsorenson) [reply shared/src/business/entities/Document.test.js]: Updated!

File: shared/src/business/entities/Document.test.js (hunk 12 above)
Comment (ericsorenson) [reply shared/src/business/entities/Document.test.js]: Updated!

File: shared/src/business/entities/Document.test.js (hunk 13 above)
Comment (ericsorenson) [reply shared/src/business/entities/Document.test.js]: Updated!

File: web-api/src/middleware/apiGatewayHelper.test.js (hunk 14 above)
Comment (ericsorenson) [reply web-api/src/middleware/apiGatewayHelper.test.js:302]: The test description was incorrect, we're updating it!

File: cypress-smoketests/integration/filings-and-qc.spec.js (hunk 6 above)
Comment (ericsorenson) [reply cypress-smoketests/integration/filings-and-qc.spec.js:28]: Yes, skipped tests are flagged by our ESLint config. Going forward, any time we skip a test we should have a comment describing why.

Comment (ericsorenson) [reply cypress-smoketests/integration/filings-and-qc.spec.js:30]: Added a comment!

File: web-api/src/middleware/apiGatewayHelper.test.js (hunk 14 above)
Comment (ericsorenson) [reply web-api/src/middleware/apiGatewayHelper.test.js:302]: Updated!

File: web-client/integration-tests/journey/petitionsClerkAddsNoticeToCase.js (hunk 15 above)
Comment (ericsorenson) [reply web-client/integration-tests/journey/petitionsClerkAddsNoticeToCase.js:46]: It was removed as part of a story, I'm adding a TODO with details.

File: cypress-smoketests/support/pages/login.js (hunk 7 above)
Comment (ericsorenson) [reply cypress-smoketests/support/pages/login.js:75]: @julialeague can you be more specific about what you're looking for?

File: cypress-smoketests/fixtures/caseMigrations.js (hunk 3 above)
Comment (ericsorenson) [reply cypress-smoketests/fixtures/caseMigrations.js:13]: I'll add a technical card to the backlog.

File: cypress-smoketests/support/pages/login.js (hunk 7 above)
Comment (julialeague) [reply cypress-smoketests/support/pages/login.js:75]: @ericsorenson I was thinking that the url as a whole could be stored in a config file and updated there as needed - what if the AWS url pattern changes or something?

**Comment by julialeague** [comment]
We can continue discussing the comments that aren't yet addressed, but I think we're at a point where any work could be added to the backlog for those. +1 to the PR in general!

File: web-client/src/presenter/actions/navigateToMessageDetailAction.test.js (hunk 16 above)
Comment (ericsorenson) [reply web-client/src/presenter/actions/navigateToMessageDetailAction.test.js]: The test description was incorrect. It has been updated!
